cpp/src/arrow/compute/exec.cc

```diff
@@ -498,15 +510,37 @@ struct NullGeneralization {
     return PERHAPS_NULL;
   }
 
+  static type Get(const ChunkedArray& chunk_array) {
+    if (chunk_array.num_chunks() == 0) {
+      return ALL_VALID;
+    }
+    if (chunk_array.null_count() == chunk_array.length()) {
```

Comment:
This does not follow the same constraint as Array above: _"Do not count the bits if they haven't been counted already"_.

I think you should instead just iterate on the chunks, such as (untested):
```c++
  static type Get(const ArraySpan& array) {
    // Do not count the bits if they haven't been counted already
    if ((arr.null_count == 0) || (arr.buffers[0].data == nullptr)) {
      return ALL_VALID;
    }
    if (arr.null_count == arr.length) {
      return ALL_NULL;
    }
    return PERHAPS_NULL;
  }

  static type Get(const ChunkedArray& chunk_array) {
    std::optional<type> current_gen;
    for (const auto& chunk : chunk_array.chunks()) {
      if (chunk->length() == 0) {
        continue;
      }
      chunk_gen = Get(*chunk);
      if (current_gen.has_value() && chunk_gen != *current_gen) {
        return PERHAPS_NULL;
      }
      current_gen = chunk_gen;
    }
    return current_gen.value_or(ALL_VALID);
  }
```

Reply:
Yes, you're right, we shouldn't count the bits in chunked_array's null generate.

In addition, I found that support `ChunkedArray` type in `NullGeneration` seems unnecessary. Whether in `ScalarExecutor` or `VectorExecutor`, chunked-array will be split into corresponding chunks for execution.
`ScalarExecutor`: chunked-array will be split into chunk in `SpanIterator`
https://github.com/apache/arrow/blob/237987ad6ab8ac95f15d6d9aa825e7ccfa2a1d99/cpp/src/arrow/compute/exec.cc#L856-L861
`VectorExecutor`: chunked-array can be executed by `SpanIterator` or their own `exec_chunked`.

 So the input parameter of `NullGeneration` will not appear in `ChunkedArray` type, i should remove the added logic for chunked array's null generation.

Reply:
Thanks for this, addressed.